controllers/imageupdateautomation_controller.go

```diff
@@ -784,15 +784,28 @@ func switchToBranch(repo *libgit2.Repository, ctx context.Context, branch string
 	}
 	defer commit.Free()
 
-	localBranch, err := repo.LookupBranch(branch, libgit2.BranchLocal)
+	localBranch, err := repo.References.Lookup(fmt.Sprintf("refs/heads/%s", branch))
 	if err != nil && !libgit2.IsErrorCode(err, libgit2.ErrorCodeNotFound) {
 		return fmt.Errorf("cannot lookup branch '%s': %w", branch, err)
 	}
 	if localBranch == nil {
-		localBranch, err = repo.CreateBranch(branch, commit, false)
-	}
-	if localBranch == nil {
-		return fmt.Errorf("cannot create local branch '%s': %w", branch, err)
+		// We could've done something like:
+		// lb, err := repo.CreateBranch()
+		// if lb != nil {
+		//	 defer lb.Free()
+		//	 localBranch = lb.Reference
+		// }
+		// But for some reasone, calling `lb.Free()` under any circumstances, causes a really
+		// nasty crash. Since not calling `lb.Free()` could result in memory leaks, we avoid
+		// assigning a variable like `lb` and get the branch reference manually.
+		_, err := repo.CreateBranch(branch, commit, false)
```

Comment:
I don't think if we just skip assigning the branch to a variable, there won't be a memory leak. Since C doesn't have GC and it's Go's responsibility to free it. Not taking the responsibility of freeing the memory would not free the memory automatically.

Reply:
We do have similar code in SC https://github.com/fluxcd/source-controller/blob/dc1037d9621b80f1639f6ed9c4bfe98db86d24b9/pkg/git/libgit2/checkout.go#L169.

Reply:
So, I managed to reproduce this running the test.
Using the reference seems to cause the crash. Let's not use the reference but at least assign the returned branch and free it.